### PR TITLE
Run `test:prepare` before `bin/rails test` commands

### PR DESCRIPTION
**Pull request by ghiculescu**
First, some background. Currently there's two ways to run Rails tests.

**`bin/rake test`**

This executes one of the rake tasks from [testing.rake](https://github.com/rails/rails/blob/main/railties/lib/rails/test_unit/testing.rake), which all call `Rails::TestUnit::Runner.rake_run`, which makes a `system` call to `rails test` ([here](https://github.com/rails/rails/blob/90a272a1de6f13711943139c4292336dbff19c7c/railties/lib/rails/test_unit/runner.rb#L34)).

But because it's a Rake task, you can also enhance it with other tasks. Most of the tasks in `testing.rake` are enhanced with `test:prepare`. This is a [placeholder](https://github.com/rails/rails/blob/90a272a1de6f13711943139c4292336dbff19c7c/railties/lib/rails/test_unit/testing.rake#L20) task, which gets hooked into by other things. For example, `cssbundling-rails` [enhances](https://github.com/rails/cssbundling-rails/blob/def749dc2dbc31f64c489438a765ce1e4315e14e/lib/tasks/cssbundling/build.rake#L15) this task to build your CSS file when you run tests. This is necessary, becuase otherwise things like `stylesheet_link_tag "tailwind"` will crash in a CI environment (because the asset does not exist).

**`bin/rails test`**

This runs the `Rails::Command::TestCommand`, which actually launches the `Rails::TestUnit::Runner` via its [`run` method](https://github.com/rails/rails/blob/90a272a1de6f13711943139c4292336dbff19c7c/railties/lib/rails/test_unit/runner.rb#L40).

As noted above, when you do `bin/rake test` you are also indirectly calling `bin/rails test`.

But because `bin/rails test` does *not* call `bin/rake test`, `rake test:prepare` is never run.

## The problem

The [Testing Rails Applications](https://guides.rubyonrails.org/testing.html#the-rails-test-runner) guide is pretty clear that you should use `bin/rails test`.

As described above, this doesn't work if you have anything that relies on `test:prepare`. Instead you have to run some other commands to properly set up your test environment. You might get lucky in development (because your CSS might get compiled during development) but you won't in CI, or production. This leads to confusing instructions like https://github.com/rails/tailwindcss-rails/pull/230 and https://github.com/rails/tailwindcss-rails/pull/134 and makes it just that little bit harder to get started and run tests all with one command.

## The solution

@dhh alluded to it [here](https://github.com/rails/tailwindcss-rails/pull/134#discussion_r786416181) - we should just run the `test:prepare` in more cases. This PR implements a few things:

- `bin/rake test` will now run `test:prepare`. Previously it did not, only subcommands (eg. `bin/rake test:all`) did.
- `bin/rails test` will now run `test:prepare`. It does this by making a `system` call to the `rake` command. Basically the inverse of how `bin/rake test` works in making a system call to the `rails` command.

That's what happens if you just call the commands with no argument. If you include a file path for a specific test:

- `bin/rake test test/models/foo_test.rb` will run `test:prepare` and run all tests. This was the behaviour before this PR. Rake tasks don't support passing arguments like this, so this has always run the full test suite.
- `bin/rails test test/models/foo_test.rb` will *not* run `test:prepare`, and will only run `foo_test.rb`. This matches the behaviour before this PR.

To summarise. Before this PR:

| Does `test:prepare` run? | `bin/rake test` | `bin/rails test` | `bin/rake test:all` |`bin/rails test:all` |
|--------------------------|----------------|---------------|-------------------|-------------------|
| No file path provided        | 🔴                      | 🔴                     | ✅                           | 🔴                           |
| File path provided             | ✅ (path is ignored) | 🔴                      | ✅                           |N/A                         |

With this PR:

| Does `test:prepare` run? | `bin/rake test` | `bin/rails test` | `bin/rake test:all` | `bin/rails test:all` |
|--------------------------|----------------|---------------|-------------------|-------------------|
| No file path provided        | ✅                      | ✅                      | ✅                           | ✅                           |
| File path provided             | ✅ (path is ignored) | 🔴                      | ✅                           | N/A                         |

Which means the "just run `bin/rails test`" advice now works in CI 👍

### CODE REVIEW

File: railties/lib/rails/commands/test/test_command.rb [hunk 1]
```diff
@@ -57,6 +60,12 @@ def generators(*)
         args.prepend("test/lib/generators")
         perform
       end
+
+      private
+        def run_prepare_task
+          success = kernel_system("rake", "test:prepare")
```
Comment (ghiculescu) [inline railties/lib/rails/commands/test/test_command.rb:66]: @etiennebarrie given your work on https://github.com/rails/rails/pull/44106 and https://github.com/rails/rails/pull/41132 I'm hoping you can help with this one 🙏  (At the very least it explains why the `test:prepare` task exists.)

Some tests are failing in CI, it looks like they are all cases where we're trying to run `rails test` on a plugin or engine which doesn't have any associated Rake tasks. So the command isn't found, and this raises.

I tried this implementation as an alternative:

```diff
diff --git a/railties/lib/rails/commands/test/test_command.rb b/railties/lib/rails/commands/test/test_command.rb
index 81be58b88f..1c00327d95 100644
--- a/railties/lib/rails/commands/test/test_command.rb
+++ b/railties/lib/rails/commands/test/test_command.rb
@@ -1,6 +1,7 @@
 # frozen_string_literal: true

 require "rails/command"
+require "rails/commands/rake/rake_command"
 require "rails/test_unit/runner"
 require "rails/test_unit/reporter"

@@ -50,7 +51,6 @@ def all(*)
         perform
       end

-      alias kernel_system system
       def system(*)
         args.prepend("test/system")
         perform
@@ -63,8 +63,7 @@ def generators(*)

       private
         def run_prepare_task
-          success = kernel_system("rake", "test:prepare")
-          success || exit(false)
+          Rails::Command::RakeCommand.perform("test:prepare")
         end
     end
   end
```

But I got the error you mention in #44106, where it was changing the environment from test to development when tests run.

Anyway, I'm just wondering if you (or anyone else reading) have any ideas how we could run the Rake task here *only* if it's defined.

**Comment by skipkayhil** [comment]
Very related: #45431 and #46098. It looks like `rspec-rails` invokes it in the rake task:

https://github.com/rspec/rspec-rails/blob/c60ff7907559653cd9d1ec1a6113bf86c9359fab/lib/rspec/rails/tasks/rspec.rake#L26

**Comment by ghiculescu** [comment]
Hmm, the issue here is if you launch your test runner *without* using rake. `bin/rails test` does not ever run a rake task as far as I can see. So that approach won't work.

**Comment by dhh** [comment]
We have to ensure that db:prepare is only run on a clean "rails test" and not on "rails test test/model/specific_case.rb". Because the setup cost of db:prepare is not reasonably for continued running of individual test cases, only when we run a larger suite.

File: railties/lib/rails/commands/test/test_command.rb (hunk 1 above)
Comment (etiennebarrie) [reply railties/lib/rails/commands/test/test_command.rb:66]: Rake has `Rake::Task.task_defined?` which you could use. I think it'd be relatively straightforward to add a `Rails::Command::RakeCommand.perform_if_exists(task)` that uses that.

Regarding the environment, it gets set in `Rails::TestUnit::Runner.parse_options`, so you need to call `run_prepare_task` after that to use the test environment (or whichever environment is asked with `--environment`).

**Comment by etiennebarrie** [comment]
In #44106 I was considering deprecating and removing `test:prepare` since it wasn't documented or used. Loading all the Rake tasks, even without taking into account the time spent in `test:prepare`, takes quite some time.

> Because the setup cost of db:prepare is not reasonably for continued running of individual test cases, only when we run a larger suite.

If it's only to be used when `bin/rails test` runs (which does not even run the whole test suite since it excludes system tests), I wonder if we wouldn't be better off pushing this to Rake only. Just like `bin/rspec` doesn't run `test:prepare` (since it has nothing to do with Rake), only `bin/rake spec` runs it, which you can then use in your CI for example.

What I'm proposing is that `bin/rake test` calls `test:prepare`, but `bin/rails test` doesn't. For individual test cases, I think it's reasonable to use a watcher process (`yarn build --watch`) which makes sure test runs stay snappy.

**Comment by ghiculescu** [comment]
> What I'm proposing is that `bin/rake test` calls `test:prepare`, but `bin/rails test` doesn't

@dhh how do you feel about this?

It's much easier to implement (which I like!), but then we'd still have two separate commands to run tests.

If we do this then we should also update the Testing in Rails guide and `-bundling-rails` gems docs to mention that this is the preferred command to run in CI. I can do all that if we're 👍 on this approach.

File: railties/lib/rails/commands/test/test_command.rb (hunk 1 above)
Comment (ghiculescu) [reply railties/lib/rails/commands/test/test_command.rb:66]: That tip about `parse_options` is what I was missing. Thank you!!

**Comment by dhh** [comment]
No, I don't think it's a good split. `bin/rake test` and `bin/rails test` should do the same thing: Run all tests (excluding the system tests), and run db:prepare before that. Ideally `bin/rake test test/models/some_test.rb" and `bin/rails test test/models/some_test.rb` should also do the same: Just run that test, not db:prepare. (`bin/rails test:models` should also run `db:prepare`).

So basically: Running the whole suite, or all tests of one category, will run db:prepare, and we accept the performance penalty. Running a single suite does not (because that's the critical quick loop during development). But bin/rake test and bin/rails test needs to be the same.

**Comment by ghiculescu** [comment]
Got the tests passing. I still need to make it only run `prepare` if no file is passed.

**Comment by ghiculescu** [comment]
I think this does what we want to it do now. @dhh could you take another look? The only oddity is that `bin/rake test path_to_test.rb` doesn't just run that path (it runs the whole suite), but that's also how it works on `main`.

File: railties/lib/rails/commands/rake/rake_command.rb [hunk 2]
```diff
@@ -21,11 +21,26 @@ def perform(task, args, config)
             if Rails.respond_to?(:root)
               rake.options.suppress_backtrace_pattern = /\A(?!#{Regexp.quote(Rails.root.to_s)})/
             end
+            unless exit_if_task_does_not_exist
+              # Check if the task does not exist, and safely abort in that case.
+              # If we don't do this, `standard_exception_handling` will exit(false) if the task doesn't exist.
+              return unless task_exists?(rake, task)
```
Comment (dhh) [inline railties/lib/rails/commands/rake/rake_command.rb:27]: This double unless is hard to parse. Go with a complete conditional:

```ruby
if exit_if_task_does_not_exist && !task_exists?(rake, task)
  return
else
  rake.standard_exception_handling { rake.top_level }
end
```

File: railties/lib/rails/commands/rake/rake_command.rb [hunk 3]
```diff
@@ -21,11 +21,26 @@ def perform(task, args, config)
...
 
         private
+          def task_exists?(rake, task)
+            name, _args = rake.parse_task_string(task)
+            begin
+              rake[name]
+              true
+            rescue Exception
+              false
+            end
```
Comment (dhh) [inline railties/lib/rails/commands/rake/rake_command.rb:41]: Don't need begin/indention:

```ruby
          def task_exists?(rake, task)
            name, _args = rake.parse_task_string(task)
             rake[name]
             true
           rescue Exception
             false
           end
```

File: railties/lib/rails/commands/test/test_command.rb [hunk 4]
```diff
@@ -57,6 +61,15 @@ def generators(*)
         args.prepend("test/lib/generators")
         perform
       end
+
+      private
+        def run_prepare_task(args)
```
Comment (dhh) [inline railties/lib/rails/commands/test/test_command.rb:66]: Use complete conditional:

```ruby
if @force_prepare || args.empty?
  Rails::Command::RakeCommand.perform("test:prepare", nil, {}, exit_if_task_does_not_exist: false)
end
```

File: railties/lib/rails/commands/rake/rake_command.rb (hunk 2 above)
Comment (ghiculescu) [reply railties/lib/rails/commands/rake/rake_command.rb]: Yeah I agree. I renamed the variable to something easier to parse.

**Comment by ghiculescu** [comment]
@dhh thanks for the fresh set of eyes, the PR is looking better for it.

File: railties/lib/rails/commands/rake/rake_command.rb [hunk 5]
```diff
@@ -21,11 +21,21 @@ def perform(task, args, config)
             if Rails.respond_to?(:root)
               rake.options.suppress_backtrace_pattern = /\A(?!#{Regexp.quote(Rails.root.to_s)})/
             end
+            return if optional && !task_exists?(rake, task)
```
Comment (dhh) [inline railties/lib/rails/commands/rake/rake_command.rb:24]: Should convert this exit wound to just a regular conditional:

```ruby
rake.standard_exception_handling { rake.top_level } unless optional && !task_exists?(rake, task)
```

File: railties/lib/rails/commands/rake/rake_command.rb [hunk 6]
```diff
@@ -21,11 +21,21 @@ def perform(task, args, config)
...
         end
 
         private
+          def task_exists?(rake, task)
+            name, _args = rake.parse_task_string(task)
+            rake[name]
+            true
+           rescue Exception
+             false
+           end
```
Comment (dhh) [inline railties/lib/rails/commands/rake/rake_command.rb:37]: Indention is off.

**Review by dhh — APPROVED** [review]

**Comment by rafaelfranca** [comment]
Is this now running `test:db:prepare`? If it is it is broken. Years ago we introduced the automatic test database preparation that still happens automatically. I get we want to run the other things that are part of `test:prepare` but we should not be running `test:db:prepare` twice before each test.

**Comment by ghiculescu** [comment]
I can look into that, does it also get called manually as well depending on test:prepare?

**Comment by rafaelfranca** [comment]
No. It is automatic by this line in the test_helper https://github.com/rails/rails/blob/e7846516f05a4c95c7f57fb86fc52a5e4915d08c/railties/lib/rails/test_help.rb#L17

**Comment by ghiculescu** [comment]
^ that calls `db:test:prepare`, so I'm assuming you meant that (not `test:db:prepare`).

`db:test:prepare` does not get invoked by `test:prepare`. I found two places that invoke `db:test:prepare`:

- the test_help line you linked
- [`rake test:db`](https://github.com/rails/rails/blob/e7846516f05a4c95c7f57fb86fc52a5e4915d08c/railties/lib/rails/test_unit/testing.rake#L28)

I wrote a test case and it passed (let me know if you want me to make a PR to add it):

```ruby
    def test_rails_test_runs_test_db_prepare
      app_file "Rakefile", <<~RUBY, "a"
        task :echo do
          puts "echo"
        end
        Rake::Task["db:test:prepare"].enhance(["echo"])
      RUBY
      create_test_file :models, "foo"
      output = run_test_command("", allow_failure: false)
      assert_equal 1, output.scan(/echo/).count # this would fail if it ran twice
      assert_includes output, "1 runs, 1 assertions, 0 failures"
    end
```

So as far as I can tell, this PR doesn't add a second call to `db:test:prepare`.

**Comment by rafaelfranca** [comment]
👍🏽. thanks for confirming.

**Comment by matt17r** [comment]
Is there an incantation to achieve the old functionality? I'd like to run all (non-system) tests without having yarn compile css and js.

I can skip `test:prepare` by running specific tests but listing all folders other than system is verbose and not future proof:
`bin/rails test test/channels/ test/controllers/ test/integration/ test/mailers/ test/models/`

In theory I can run everything in `test/` and exclude `test/system` using `--exclude` but I haven't been able to figure out the right syntax using strings or regex:
`bin/rails test test/ --exclude test/system` **does not work**

**Comment by ghiculescu** [comment]
No, there isn't. Is the `test:prepare` step pretty slow for you?

Try this:

```
DEFAULT_TEST_EXCLUDE="test/{system,dummy}/**/*_test.rb" bin/rails test test/**/*_test.rb
```

It's based on [this method](https://github.com/rails/rails/blob/632bcb391840bacc77556eee6707178ddf293b99/railties/lib/rails/test_unit/runner.rb#L101).

**Comment by ghiculescu** [comment]
> I also tried the inverse, `DEFAULT_TEST="test/{channels,controllers,integration,mailers,models}/**/*_test.rb" bin/rails test`, but it has the same problem.

Can you provide both ENVs?

`DEFAULT_TEST="test/{channels,controllers,integration,mailers,models}/**/*_test.rb" DEFAULT_TEST_EXCLUDE="test/system/**/*_test.rb" bin/rails test`

Otherwise, your solution looks like it'll work too. Hopefully you aren't adding new folders too often 🙃 


> I'm still not sure why `bin/rails test test/ --exclude test/system` doesn't work. The help text from `-h` doesn't make it any clearer ("Exclude /regexp/ or string from run.") and [the guide doesn't add any info](https://edgeguides.rubyonrails.org/testing.html#the-rails-test-runner).

My read of the help guide is that it's an argument `minitest` supports to exclude specific tests. You can supply the test's name or a regex to match the test name. It's not for files.

**Comment by matt17r** [comment]
> Can you provide both ENVs?

Yes but it will have the same problem. If there's no arguments to `test` then `test:prepare` will run.

> My read of the help guide is that it's an argument `minitest` supports to exclude specific tests. You can supply the test's name or a regex to match the test name. It's not for files or paths.

You are 100% correct. I was assuming paths would work but it's minitest names. I've investigated how that works and had a go at [improving the guide to cover how that works in Rails](https://github.com/rails/rails/pull/49964).

Thanks for taking the time to respond @ghiculescu! Now I know the rationale for the change I don't expect things to change based on one person's feedback but I'll leave these comments here for posterity. If someone else ever raises a bug wanting to change how `bin/rails test` works or add a flag to skip asset preparation they can link back to these comments 😄

**Comment by searls** [comment]
Appreciate the conversation in this thread. 

It kinda has me thinking it'd be a fun afternoon project to write a script that dynamically cleared the `test:prepare` task if no files had changed under `app/assets` and the other various relevant locations since the most recent test run. That way, someone could run `bin/rails test` and get the certainty of knowing `test:prepare` will run, but then run it repeatedly and not redundantly pay the tax of whatever asset tasks exist. (For me, the yarn install alone is an order of magnitude slower than my non-system test suite)

In the meantime, I just wrote this `script/unit_test` snippet for myself to get fast feedback when I know it's safe:

```
#!/usr/bin/env bash

find test -type f -name "*_test.rb" | grep -v "test/system" | xargs bin/rails test
```

**Comment by ghiculescu** [comment]
> For me, the yarn install alone is an order of magnitude slower than my non-system test suite

Even when there’s no changes to your lockfile? That seems wrong.

**Comment by erlingur** [comment]
This change broke my test suite for some reason. For some reason, after updating from Rails 7.0 to 7.1 I suddenly got a really strange error when my Controller tests ran and it tried to render a view:

"arguments passed to url_for can't be handled. Please require routes or provide your own implementation"

The (very) annoying thing was that this only happened when I ran the whole test suite. Running the individual test made everything pass again. I noticed that when I ran the whole test suite the `url_for` method that `link_to` was calling was the one defined in the [url_helper.rb](https://github.com/rails/rails/blob/main/actionview/lib/action_view/helpers/url_helper.rb) file, but when when I specified the controller test, like `rails test test/controller/posts_controller_test.rb`, then the `link_to` in my views was calling the `url_for` method defined in the [routing_url_for.rb](https://github.com/rails/rails/blob/main/actionview/lib/action_view/routing_url_for.rb) file.

After a LOT of headscratching I decided do just bisect the whole 7.1 stable branch until I finally found the commit that broke this, which was the one in this PR.

If I comment out https://github.com/rails/rails/blob/965b8c372ac0e19021f92cf798095af6046ab140/railties/lib/rails/commands/test/test_command.rb#L32

Then my whole test suite runs completely fine. I'm still trying to figure out exactly what's happening here, I don't fully understand it. I've tried to reproduce this in a new project but I was unable to. It's probably something to do with my environment, maybe some gem is hooking into the `test:prepare` task and changing something. I figured I would write this up now while I'm working on this in the hopes that maybe someone could provide some more insight.

**Comment by erlingur** [comment]
I've spent so much time trying to figure this out and I have FINALLY fixed it so I thought I would post what the issue was. This isn't a problem in Rails but rather (as I suspected) a gem I was using,  CableReady v5.0.0-pre10. I suspect that the issue was that they had an `include Rails.application.routes.url_helpers` in one of their Rake tasks which was causing this issue. This was fixed in this pr: https://github.com/stimulusreflex/cable_ready/pull/267

After updating to the latest version of the CableReady gem life is good again! Posting this in case anyone else has a similar issue and this helps them.

**Comment by ghiculescu** [comment]
Thanks for sharing @erlingur and sorry you had to spend so long debugging it!

**Comment by erlingur** [comment]
> Thanks for sharing @erlingur and sorry you had to spend so long debugging it!

No worries, it was a very good learning experience if nothing else 😄 Just happy I figured it out. Thanks @ghiculescu 😄
